velox/dwio/common/tests/TypeTests.cpp

```diff
@@ -43,20 +58,33 @@ TEST(TestType, selectedType) {
   EXPECT_EQ(0, typeWithId->id());
   EXPECT_EQ(11, typeWithId->maxId());
 
+  auto copySelector = [](size_t index) { return true; };
+  auto typeWithIdCopy = buildSelectedType(typeWithId, copySelector);
+  assertEqualTypeWithId(typeWithId, typeWithIdCopy);
+
   std::vector<bool> selected(12);
   selected[0] = true;
   selected[2] = true;
   auto selector = [&selected](size_t index) { return selected[index]; };
   auto cutType = buildSelectedType(typeWithId, selector);
+  assertEqualTypeWithId(typeWithIdCopy, typeWithId);
   EXPECT_STREQ(
       "struct<col1:smallint>",
       HiveTypeSerializer::serialize(cutType->type()).c_str());
   EXPECT_EQ(0, cutType->id());
   EXPECT_EQ(11, cutType->maxId());
   EXPECT_EQ(2, cutType->childAt(0)->maxId());
 
+  // The following two lines verify that the original type tree's children are
+  // not re-parented by the buildSelectedType method when copying. If it is
+  // re-parented, then this test would crash with SIGSEGV. The return type is
+  // deliberately ignored so the copied type will be deallocated upon return.
+  buildSelectedType(typeWithId, selector);
+  EXPECT_EQ(typeWithId->childAt(1)->parent()->type()->kind(), TypeKind::ROW);
```

Comment:
@nmahadevuni Instead of just checking a single unslected child, I think we should check every node's children's parents are itself. Can you please create another function assertValidTypeWithId() function and call it on both the cutType and original typeWithId after every selection is done?

Reply:
This is the selected child, the selected[2]=true translates to child at index 1 in the type tree. This test crashes without the fix, so I kept this to validate that parent node.